scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,82 @@
+
+A Scanner instance runs the Vulnerability Updater (aka. Updater) to fetch the bundle and use ClairCore to parse the schema and update the vulnerability database.  Suppose ClairCore changed the schema in a backward-incompatible manner, and that change is incorporate in the Scanner repository (a ClairCore version bump).  In that scenario, the next Exporter execution will update the bundle with the new schema.  Then, existing Scanner instances won't be able to consume the data.  A Scanner version bump to a version of ClairCore supporting the new schema would be required.
+
+To address that scenario, Scanner created one bundle per Stackrox release.  Each bundle is created with an Exporter running the same ClairCore version used to build the release, ensuring compatibility.  Bundles are served at `https://definitions.stackrox.io/v4/V/vulnerabilities.zip` where `V` will be a Stackrox release version such as "4.5.0".  Scanner uses its release version tagged during build time to determine which version to fetch from.  The versions to build bundles for are taken from a file named `scanner/updater/version/RELEASE_VERSION`.
+
+For its purpose, this approach needs to be revised.
+
+1. The vulnerability schema rarely changes, but maintaining multiple bundles is expensive.  It requires multiple GitHub workflow runs, which increases the chances of failure and pressure on the security sources' endpoints.
+2. It isn't easy to backport fixes and updates to previous releases.  This requires patch releases, which trigger the release workflow.  Improving the release process is noisy, convoluted, and complex.
+3. The offline mode's "latest bundle" needs to exist as a complex solution to help users transition to the versioned bundle.  It contains all versions in a single bundle.  But one bundle per release makes the latest bundle large and, most of the time, essentially carries the same data since the schema is the same.
+
+## Decision
+
+We will implement a versioning stream for vulnerability bundles, bumped when we determine that ClairCore's schema has changed.
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current Stackrox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
+
+1.  The schema hasn't changed, so v1 now tracks the new release (`heads/4.4.6`).
+2.  The schema has changed, so v1 stays at current release (`heads/4.4.5`), and a new version, v2, is created to track the next release (`heads/4.4.6`).
```

Comment:
Thoughts on using the ref directly instead of an arbitrary version?  In the example that would yield bundle versions `4.4.5` and `4.4.6`, therefore we would know immediately the min version of ACS that supports the provided bundle vs. having to do a translation.

That may simplify the rest of the approach defined below.

Reply:
That can be confusing. Let's transition happened at `4.4.5`, we have two bundles: `4.4.5` (schema v1) and `4.4.6` (schema v2). All subsequent versions for a while (4.5.0, 4.6.0, etc) would use `4.4.6` and that would raise questions. A separate mapping is cleaner, and easy to spot because we have one file on the repo (and hard-coded in the binary) of which bundle stream a given Scanner build needs.